File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import Optional
import os
import re


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts

# Write your first then lower case function here
def first_then_lower_case(strings: list[str], predicate: Callable[[str], bool]) -> Optional[str]:
  for idx in range(len(strings)):
    if strings[idx] != None and predicate(strings[idx]):
      return strings[idx].lower()
  return None

# Write your powers generator here
def powers_generator(*, base: int, limit: int):
    power: int = 1
    while power <= limit:
        yield power
        power = power * base

# Write your say function here
def say(text: Optional[str] = None) -> Callable | str:
    ''' 
    Inner function chains the strings together to create final concatenated sentence. 

    String concatenation guide to use .join method: https://www.digitalocean.com/community/tutorials/python-string-concatenation
    '''
    sentence: list[str] = []

    def final_sentence(next_text: Optional[str] = None) -> Callable | str:
        if next_text is None:
            return " ".join(sentence)
        else:
            sentence.append(next_text)

        return final_sentence
    
    if text is None:
        return ""
    
    sentence.append(text)

    return final_sentence

# Write your line count function here
def meaningful_line_count(path: str) -> int:
    path = os.path.join(os.path.dirname(__file__), "..", path)

    try:
        f = open(path, "r")
    except:
        raise FileNotFoundError("No such file")
    count = 0
    for line in f.readlines():
        newLine = line.strip()
        if not (len(newLine) == 0 or newLine[0] == "#"):
            count += 1
    f.close()
    return count

# Write your Quaternion class here
@dataclass (frozen = True)
class Quaternion():
    a: int
    b: int
    c: int
    d: int

    def __add__(self, q):
        return Quaternion(self.a + q.a, self.b + q.b, self.c + q.c, self.d + q.d)
    
    def __mul__(self, q):
        return Quaternion(((self.a * q.a) - (self.b * q.b) - (self.c * q.c) - (self.d * q.d)),
                          ((self.a * q.b) + (self.b * q.a) + (self.c * q.d) - (self.d * q.c)),
                          ((self.a * q.c) - (self.b * q.d) + (self.c * q.a) + (self.d * q.b)),
                          ((self.a * q.d) + (self.b * q.c) - (self.c * q.b) + (self.d * q.a)))
    
    @property 
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)

    @property
    def conjugate(self):
        return Quaternion(self.a, (-1 * self.b), (-1 * self.c), (-1 * self.d))

    def __eq__(self, q):
        return self.a == q.a and self.b == q.b and self.c == q.c and self.d == q.d
    
    def __str__(self):
        output = ""

        def get_num_remove_digit(coefficient, basis_vector):
            coefficient_formatted = ""
            if coefficient > 1 and basis_vector != "":
                coefficient_formatted = "+" + str(coefficient)
            elif coefficient == 1 and basis_vector != "":
                coefficient_formatted = "+"
            elif coefficient == -1 and basis_vector != "":
                coefficient_formatted = "-"
            else:
                # number formatted is the number itself if number negative or if basis_vector == ""
                coefficient_formatted = str(coefficient)
            if coefficient != 0:
                return coefficient_formatted + basis_vector
            return ""
        
        output = output + get_num_remove_digit(self.a, "")
        output = output + get_num_remove_digit(self.b, "i")
        output = output + get_num_remove_digit(self.c, "j")
        output = output + get_num_remove_digit(self.d, "k")
        if output != "":
            # remove leading "+" and all whitespace from output
            output = re.sub(r"^\+", "", output)
            output = output.strip()

        return output or "0"

File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
  function first_then_lower_case(a, p)
    for _, i in ipairs(a) do
      if p(i) then
        return string.lower(i)        
      end
    end
    return nil
  end

-- Write your powers generator here
function powers_generator(base, limit)
  local total = 1
  -- create coroutine
  return coroutine.create(function()
    while total <= limit do
      -- pause coroutine and return total
      coroutine.yield(total)
      total = total * base
    end
  -- need to close coroutine with parenthesis
  end)
end

-- Write your say function here
  function say(sentence)
    local words = {}
    -- concatenation helper function 
    local function combine(word)
      -- instead of for loop, return the function to itself so that itll run again  
      if word then
            table.insert(words, word)
            return combine 
        else
            return table.concat(words, " ")
        end
    end

    -- start the chain of words
    if sentence then
        table.insert(words, sentence)
        return combine
    else
        return ""
    end
end

-- Write your line count function here
-- open file, if file can't be open throw an error, if it can be openned iterate over each line in the file, get rid of whitespace, if not empty add it
  function meaningful_line_count(file)
      -- Attempt to open the file for reading
      local f, err = io.open(file, "r")
      local counter = 0
      if not f then
          error("Can't open file: " .. err)
      end
      for line in f:lines() do
        --whitespace strip
          line = line:match("^%s*(.-)%s*$")
          if line ~= "" and not line:find("^#") then
              counter = counter + 1
          end
      end
      f:close()
      return counter
  end

-- Write your Quaternion table here

local input = io.read() 
Quaternion = {}
for number in string.gmatch(input, "[^%s]+") do
    table.insert(Quaternion, number)
end

-- function tostring()
--   local output = ""
--   local var = {"", "i", "j", "k"}
--   local counter = 1
--   for key,value in next, Quaternion, nil do 
--     print(key)
--     output = (output.." "..value.." "..var[counter].." ")
--     counter = counter + 1
--   end
--   return output
-- end

function add()
  
end

-- 000000000000000000000000

Quaternion = {}

Quaternion = (function (class)
  class.new = function (a, b, c, d)
    return setmetatable({a = a, b = b, c = c, d = d}, {
      __index = {
        magnitude = function(self)
          return math.sqrt(self.b * self.b + self.c * self.c)
        end,

        normalized = function(self)
          local m = self:magnitude()
          return class.new(self.b / m, self.c / m)
        end,

        conjugate = function(self)
          return class.new(self.a, self.b*(-1), self.c*(-1), self.d*(-1))
        end,

        coefficients = function(self)
          return {self.a, self.b, self.c, self.d}
        end

      },
      __add = function(self, v)
        return class.new(self.a + v.a, self.b + v.b, self.c + v.c, self.d + v.d)
      end,

      __sub = function(self, v)
        return class.new(self.a - v.b, self.c - v.d)
      end,

      __mul = function(self, v)
        aProduct = (self.a * v.a) - (self.b * v.b) - (self.c * v.c) - (self.d * v.d)
        bProduct = (self.a * v.b) + (self.b * v.a) + (self.c * v.d) - (self.d * v.c)
        cProduct = (self.a * v.c) - (self.b * v.d) + (self.c * v.a) + (self.d * v.b)
        dProduct = (self.a * v.d) + (self.b * v.c) - (self.c * v.b) + (self.d * v.a)
        return class.new(aProduct, bProduct, cProduct, dProduct)
      end,

      __eq = function(self, v)
        return self.a == v.a and self.b == v.b and self.c == v.c and self.d == v.d
      end,

      __tostring = function(self)
        local function formatComponent(coefficient, vector)
          coefficient = coefficient + .0
          local str = tostring(coefficient)
          if str == '0.0' then
            return ''
          elseif str == '1.0' then
            return vector
          elseif str == '-1.0' then
            return '-' .. vector
          else
            return str ..vector
          end
        end

        self.a = self.a + .0

        if tostring(self.a) == '0.0' then
          nString = ''
        else
          nString = tostring(self.a)
        end

        local iString = formatComponent(self.b, 'i')
        local jString = formatComponent(self.c, 'j')
        local kString = formatComponent(self.d, 'k')

        returnString = nString .. '+' .. iString .. '+' .. jString .. '+' .. kString

        returnString = returnString:gsub('%++', '+')
        returnString = returnString:gsub('%+%-', '-')

        if returnString == '' then
          return '0'
        end

        if string.sub(returnString, -1) == '+' then
          returnString = string.sub(returnString, 1, -2)
        end

        if string.sub(returnString, 1, 1) == '+' then
          returnString = string.sub(returnString, 2)
        end     

        if returnString == '' then
          return '0'
        end
  
        return returnString
      end
    })
    end
  return class
end)({})